File: ruby/1-25/problem_7.rb
=begin
                                10001st prime

    By  listing  the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see
    that the 6th prime is 13.
    What is the 10 001st prime number?

    note:
    "An  array with 500 million elements is 2 GiBytes in size, which – depending
    on  the  specific OS you are using – is typically the maximum that a process
    can address. In other words: your array is bigger than your address space."
    from: http://stackoverflow.com/questions/3687517/array-size-too-big-ruby/3688388#3688388
=end

#!/usr/bin/ruby

def sieve( limit )
    max_length = 500_000_000
    primes = []
    numbers = Array.new( max_length ) { | i | true }
    numbers[ 0 ] = false
    numbers[ 1 ] = false

    max = ( Math.sqrt( max_length ) ).to_i
    for i in ( 2..max ) do
        if numbers[ i ] then
            ( i**2..max_length ).step( i ) do | j |
                numbers[ j ] = false
            end
        end
    end

    primes = numbers.each_index.select { | i | true == numbers[ i ]  }
    return primes[ limit-1 ]
end

puts sieve( 10_001 )

File: src/python/1-25/problem_7.py
"""
                                10001st prime

    By  listing  the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see
    that the 6th prime is 13.

    What is the 10 001st prime number?

                                Answer: 104743
"""

import math

def sieve(pos):
    """Given pos this function returns the determinated prime"""
    #   limit was a totally abritary number that I assigned
    limit = 5000000
    numbers = [True for _ in range(1, limit)]
    numbers[0] = numbers[1] = False

    for index_i in range(2, int(math.sqrt(limit))):
        if numbers[index_i]:
            for index_ii in range(index_i**2, limit, index_i):
                numbers[index_ii] = False

    return [x for x, y in enumerate(numbers) if y][pos-1]

print(sieve(10001))
